File: go/internal/cli/maintenance.go
package cli

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"time"

	"github.com/nexusriot/ai-houkai/internal/maintenance"
	reflectpkg "github.com/nexusriot/ai-houkai/internal/reflect"
	"github.com/spf13/cobra"
)

// maintCfg builds a maintenance.Config from the resolved CLI config.
func maintCfg(cfg Config) maintenance.Config {
	interval := time.Duration(cfg.Maintenance.IntervalSecs) * time.Second
	if interval <= 0 {
		interval = time.Hour
	}
	summ, err := reflectpkg.BuildSummarizer(cfg.Summarizer, true)
	if err != nil {
		summ = nil // reflect.New falls back to the extractive summarizer
	}
	consolidate := reflectpkg.ConsolidateNone
	if cfg.Maintenance.Consolidate {
		consolidate = reflectpkg.ConsolidateSoft
	}
	return maintenance.Config{
		Interval:        interval,
		DecayRate:       cfg.Maintenance.Decay.DecayRate,
		MinScore:        cfg.Maintenance.Decay.MinScore,
		Reflect:         cfg.Maintenance.Reflect,
		Consolidate:     consolidate,
		DecayEvery:      float64(cfg.Maintenance.DecayEverySecs),
		ReflectEvery:    float64(cfg.Maintenance.ReflectEverySecs),
		FrequencyWeight: cfg.Maintenance.Decay.FrequencyWeight,
		Summarizer:      summ,
	}
}

// MaintenanceRuntime resolves the maintenance.Config and state path for cfg —
// used by the MCP entry point to wire the schedule-gated maintenance_tick.
func MaintenanceRuntime(cfg Config) (maintenance.Config, string) {
	statePath, _, _ := cfg.MaintPaths()
	return maintCfg(cfg), statePath
}

// tickSummary renders a TickResult as a one-line human summary, mirroring
// Python's TickResult.summary(): gated-out jobs are reported as skipped.
func tickSummary(res maintenance.TickResult) string {
	var parts []string
	if res.RanDecay {
		parts = append(parts, fmt.Sprintf("decay pruned %d", res.Pruned))
	}
	if res.RanReflect {
		parts = append(parts, fmt.Sprintf("reflect created %d", res.Reflected))
	}
	if len(parts) == 0 {
		return "nothing to do (jobs not due yet)"
	}
	return strings.Join(parts, " | ")
}

func fmtTS(ts float64) string {
	if ts <= 0 {
		return "never"
	}
	return time.Unix(int64(ts), 0).Format("2006-01-02 15:04:05")
}

func newMaintenanceCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "maintenance",
		Short: "Scheduled decay + reflection daemon (tick|run|start|stop|status)",
	}
	cmd.AddCommand(
		newMaintTickCmd(),
		newMaintRunCmd(),
		newMaintStartCmd(),
		newMaintStopCmd(),
		newMaintStatusCmd(),
	)
	return cmd
}

func newMaintTickCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "tick",
		Short: "Run one maintenance tick synchronously (cron-friendly)",
		RunE: func(cmd *cobra.Command, _ []string) error {
			store := storeFromCtx(cmd.Context())
			cfg := cfgFromCtx(cmd.Context())
			statePath, _, _ := cfg.MaintPaths()
			res := maintenance.Tick(cmd.Context(), store, store, maintCfg(cfg), statePath, float64(time.Now().Unix()))
			fmt.Printf("Tick complete: %s\n", tickSummary(res))
			return res.Err
		},
	}
}

func newMaintRunCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "run",
		Short: "Run the maintenance scheduler in the foreground (Ctrl-C to stop)",
		RunE: func(cmd *cobra.Command, _ []string) error {
			store := storeFromCtx(cmd.Context())
			cfg := cfgFromCtx(cmd.Context())
			statePath, _, _ := cfg.MaintPaths()
			mc := maintCfg(cfg)

			ctx, stop := signal.NotifyContext(cmd.Context(), syscall.SIGTERM, syscall.SIGINT)
			defer stop()

			log.Printf("maintenance: running every %s (reflect=%v)", mc.Interval, mc.Reflect)
			runOne := func() {
				res := maintenance.Tick(ctx, store, store, mc, statePath, float64(time.Now().Unix()))
				if res.Err != nil {
					log.Printf("maintenance tick error: %v", res.Err)
				} else {
					log.Printf("maintenance tick: %s", tickSummary(res))
				}
			}
			runOne() // one immediate pass on startup
			ticker := time.NewTicker(mc.Interval)
			defer ticker.Stop()
			for {
				select {
				case <-ctx.Done():
					log.Print("maintenance: stopping")
					return nil
				case <-ticker.C:
					runOne()
				}
			}
		},
	}
}

func newMaintStartCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "start",
		Short: "Detach the maintenance daemon into the background",
		RunE: func(cmd *cobra.Command, _ []string) error {
			cfg := cfgFromCtx(cmd.Context())
			_, pidPath, logPath := cfg.MaintPaths()
			if maintenance.IsAlive(pidPath) {
				fmt.Printf("Daemon is already running (pid %d).\n", maintenance.ReadPid(pidPath))
				return nil
			}
			exe, err := os.Executable()
			if err != nil {
				return err
			}
			pid, err := maintenance.SpawnDetached(exe, cfg.StorePath, cfg.Collection, logPath, pidPath)
			if err != nil {
				return err
			}
			fmt.Printf("Maintenance daemon started (pid %d).\n", pid)
			fmt.Printf("Logs → %s\nStop → houkai maintenance stop\n", logPath)
			return nil
		},
	}
}

func newMaintStopCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "stop",
		Short: "Send SIGTERM to the background daemon",
		RunE: func(cmd *cobra.Command, _ []string) error {
			cfg := cfgFromCtx(cmd.Context())
			_, pidPath, _ := cfg.MaintPaths()
			if !maintenance.IsAlive(pidPath) {
				if pid := maintenance.ReadPid(pidPath); pid > 0 {
					fmt.Printf("Daemon (pid %d) is not running. Cleaning up stale pid file.\n", pid)
					maintenance.RemovePid(pidPath)
				} else {
					fmt.Println("No daemon is running.")
				}
				return nil
			}
			pid := maintenance.ReadPid(pidPath)
			if maintenance.StopDaemon(pidPath) {
				fmt.Printf("SIGTERM sent to daemon (pid %d).\n", pid)
				maintenance.RemovePid(pidPath)
			} else {
				fmt.Println("Failed to send signal — daemon may have already exited.")
			}
			return nil
		},
	}
}

func newMaintStatusCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "status",
		Short: "Show daemon state, last runs, and next schedules",
		RunE: func(cmd *cobra.Command, _ []string) error {
			cfg := cfgFromCtx(cmd.Context())
			statePath, pidPath, logPath := cfg.MaintPaths()
			state := maintenance.LoadState(statePath)
			pid := maintenance.ReadPid(pidPath)
			alive := maintenance.IsAlive(pidPath)

			daemon := "stopped"
			if alive {
				daemon = fmt.Sprintf("running (pid %d)", pid)
			} else if pid > 0 {
				daemon = "stopped (stale pid file)"
			}

			line := "────────────────────────────────────────────────────────"
			fmt.Println(line)
			fmt.Printf("  Daemon:          %s\n", daemon)
			fmt.Println(line)
			fmt.Printf("  Last decay:      %s\n", fmtTS(state.LastDecayAt))
			fmt.Printf("  Last reflect:    %s\n", fmtTS(state.LastReflectAt))
			fmt.Printf("  Total decayed:   %d\n", state.TotalDecayed)
			fmt.Printf("  Total reflected: %d\n", state.TotalReflected)
			fmt.Println(line)
			fmt.Printf("  Interval:        %ds\n", cfg.Maintenance.IntervalSecs)
			fmt.Printf("  Reflect:         %v\n", cfg.Maintenance.Reflect)
			fmt.Printf("  State file:      %s\n", statePath)
			fmt.Printf("  Log file:        %s\n", logPath)
			reinf := "off"
			if cfg.Maintenance.Decay.FrequencyWeight != 0 {
				reinf = fmt.Sprintf("on (frequency_weight=%g)", cfg.Maintenance.Decay.FrequencyWeight)
			}
			fmt.Printf("  Reinforcement:   %s\n", reinf)
			return nil
		},
	}
}

File: ai_houkai/cli/commands/maintenance.py
"""Maintenance sub-commands — scheduled decay + reflection daemon.

Commands
    houkai maintenance tick     Run one tick synchronously (cron-friendly).
    houkai maintenance run      Foreground blocking loop (Ctrl-C to exit).
    houkai maintenance start    Detach daemon into the background.
    houkai maintenance stop     Send SIGTERM to the background daemon.
    houkai maintenance status   Show daemon state, last runs, and next schedules.
"""

from __future__ import annotations

import logging
import signal
import threading
import time
from datetime import datetime

import typer

from ai_houkai.cli.config import load_maintenance
from ai_houkai.maintenance.daemon import (
    get_pid,
    is_alive,
    remove_pid,
    spawn_detached,
    stop_daemon,
)
from ai_houkai.maintenance.durations import format_duration
from ai_houkai.maintenance.scheduler import MaintenanceScheduler
from ai_houkai.maintenance.state import MaintenanceState
from ai_houkai.memory_system.summarizers import build_summarizer

maintenance_app = typer.Typer(
    name="maintenance",
    help="Scheduled decay + reflection daemon.",
    no_args_is_help=True,
)


def _make_scheduler(store, mcfg) -> MaintenanceScheduler:
    try:
        summarizer = build_summarizer(mcfg.summarizer)
    except (ValueError, ImportError) as e:
        typer.echo(f"Error in [maintenance.reflect].summarizer: {e}", err=True)
        raise typer.Exit(1)
    return MaintenanceScheduler(
        store=store,
        decay_every=mcfg.decay_every,
        reflect_every=mcfg.reflect_every,
        purge_every=mcfg.purge_every,
        trash_ttl_days=mcfg.trash_ttl_days,
        tick_interval=mcfg.tick_interval,
        state_path=mcfg.state_path,
        decay_rate=mcfg.decay_rate,
        min_score=mcfg.min_score,
        protect_types=mcfg.protect_types,
        frequency_weight=mcfg.frequency_weight,
        min_cluster_size=mcfg.min_cluster_size,
        reflect_apply=mcfg.reflect_apply,
        reflect_consolidate=mcfg.reflect_consolidate,
        summarizer=summarizer,
    )


_DISABLED_MSG = (
    "Maintenance is disabled ([maintenance].enabled = false in "
    "~/.config/ai_houkai/config.toml) — nothing to do."
)


def _fmt_ts(ts: float | None) -> str:
    if ts is None:
        return "never"
    return datetime.fromtimestamp(ts).strftime("%Y-%m-%d %H:%M:%S")


def _fmt_next(last_at: float | None, interval: int | None) -> str:
    if interval is None:
        return "disabled"
    if last_at is None:
        return "now (never ran)"
    next_ts = last_at + interval
    diff = next_ts - time.time()
    direction = "in" if diff > 0 else "overdue by"
    return f"{_fmt_ts(next_ts)}  ({direction} {format_duration(abs(diff))})"


@maintenance_app.command("tick")
def tick_cmd(ctx: typer.Context) -> None:
    """Run one maintenance tick synchronously and print results.

    Suitable for use from cron:  houkai maintenance tick
    """
    store = ctx.obj["store"]
    mcfg = load_maintenance()
    if not mcfg.enabled:
        typer.echo(_DISABLED_MSG)
        return
    sched = _make_scheduler(store, mcfg)

    typer.echo("Running maintenance tick…")
    result = sched.tick()
    typer.echo(f"Result: {result.summary()}")

    if result.decay_error:
        typer.echo(f"  Decay error: {result.decay_error}", err=True)
    if result.reflect_error:
        typer.echo(f"  Reflect error: {result.reflect_error}", err=True)

    if result.decay_error or result.reflect_error:
        raise typer.Exit(1)


@maintenance_app.command("run")
def run_cmd(ctx: typer.Context) -> None:
    """Run the maintenance scheduler in the foreground (Ctrl-C to stop).

    Used internally by 'start'; also useful for testing before daemonizing.
    """
    store = ctx.obj["store"]
    mcfg = load_maintenance()
    if not mcfg.enabled:
        typer.echo(_DISABLED_MSG)
        return

    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s  %(levelname)-8s  %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    stop = threading.Event()

    def _handle_term(signum, frame):
        logging.getLogger(__name__).info("Received signal %d — stopping.", signum)
        stop.set()

    signal.signal(signal.SIGTERM, _handle_term)
    signal.signal(signal.SIGINT, _handle_term)

    sched = _make_scheduler(store, mcfg)
    sched.run_forever(stop)


@maintenance_app.command("start")
def start_cmd(ctx: typer.Context) -> None:
    """Detach the maintenance daemon into the background."""
    cfg = ctx.obj["config"]
    mcfg = load_maintenance()
    if not mcfg.enabled:
        typer.echo(_DISABLED_MSG, err=True)
        raise typer.Exit(1)

    if is_alive(mcfg.pid_path):
        pid = get_pid(mcfg.pid_path)
        typer.echo(f"Daemon is already running (pid {pid}).")
        raise typer.Exit(1)

    pid = spawn_detached(
        store_path=cfg.store_path,
        collection=cfg.collection,
        log_path=mcfg.log_path,
        pid_path=mcfg.pid_path,
    )
    typer.echo(f"Maintenance daemon started (pid {pid}).")
    typer.echo(f"Logs → {mcfg.log_path}")
    typer.echo(f"Stop → houkai maintenance stop")


@maintenance_app.command("stop")
def stop_cmd(ctx: typer.Context) -> None:
    """Send SIGTERM to the background daemon."""
    mcfg = load_maintenance()

    if not is_alive(mcfg.pid_path):
        pid = get_pid(mcfg.pid_path)
        if pid:
            typer.echo(f"Daemon (pid {pid}) is not running. Cleaning up stale pid file.")
            remove_pid(mcfg.pid_path)
        else:
            typer.echo("No daemon is running.")
        return

    pid = get_pid(mcfg.pid_path)
    sent = stop_daemon(mcfg.pid_path)
    if sent:
        typer.echo(f"SIGTERM sent to daemon (pid {pid}).")
        remove_pid(mcfg.pid_path)
    else:
        typer.echo("Failed to send signal — daemon may have already exited.")


@maintenance_app.command("status")
def status_cmd(ctx: typer.Context) -> None:
    """Show daemon state, last run times, and next scheduled runs."""
    mcfg = load_maintenance()
    state = MaintenanceState.load(mcfg.state_path)

    pid = get_pid(mcfg.pid_path)
    alive = is_alive(mcfg.pid_path)
    daemon_line = (
        f"running (pid {pid})" if alive
        else ("stopped (stale pid file)" if pid else "stopped")
    )

    width = 60
    typer.echo("─" * width)
    if not mcfg.enabled:
        typer.echo("  Enabled:        false ([maintenance].enabled — "
                   "tick/run/start are no-ops)")
    typer.echo(f"  Daemon:         {daemon_line}")
    typer.echo("─" * width)
    typer.echo(f"  Last decay:     {_fmt_ts(state.last_decay_at)}")
    typer.echo(f"  Last reflect:   {_fmt_ts(state.last_reflect_at)}")
    typer.echo(f"  Total decayed:  {state.total_decayed}")
    typer.echo(f"  Total reflected:{state.total_reflected}")
    typer.echo("─" * width)
    typer.echo(f"  Next decay:     {_fmt_next(state.last_decay_at, mcfg.decay_every)}")
    typer.echo(f"  Next reflect:   {_fmt_next(state.last_reflect_at, mcfg.reflect_every)}")
    typer.echo("─" * width)
    typer.echo(f"  State file:     {mcfg.state_path}")
    typer.echo(f"  Log file:       {mcfg.log_path}")
    typer.echo(f"  reflect_apply:  {mcfg.reflect_apply}")
    consolidate = {False: "none", True: "soft"}.get(
        mcfg.reflect_consolidate, mcfg.reflect_consolidate)
    typer.echo(f"  consolidate:    {consolidate}")
    typer.echo(f"  summarizer:     {mcfg.summarizer or 'extractive (built-in)'}")
    typer.echo(
        f"  reinforcement:  "
        + (f"on (frequency_weight={mcfg.frequency_weight})"
           if mcfg.frequency_weight else "off")
    )
